backend/mora/service/address_handler/base.py

```diff
@@ -0,0 +1,148 @@
+#
+# Copyright (c) 2017-2018, Magenta ApS
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+import abc
+import inspect
+
+from ... import exceptions
+from ... import mapping
+from ... import util
+
+ADDRESS_HANDLERS = {}
+
+
+class _AddressHandlerMeta(abc.ABCMeta):
+    """Metaclass for automatically registering handlers."""
+
+    @staticmethod
+    def __new__(mcls, name, bases, namespace):
+        cls = super().__new__(mcls, name, bases, namespace)
+
+        cls._register()
+
+        return cls
+
+
+class AddressHandler(metaclass=_AddressHandlerMeta):
+    __slots__ = 'scope', 'prefix', '_value'
```

Comment:
Note that this adds very little unless all the subclasses also have a `__slots__` declaration.

Reply:
I use it simply to declare the properties available on the class. What's the alternative, declaring `scope = None` etc. for every property?

Reply:
> I use it simply to declare the properties available on the class. What's the alternative, declaring `scope = None` etc. for every property?

That's more explicit, actually, given that `__slots__` does not assign a default value. Given:

```python
class X:
    __slots__ = ('x', 'y')

    def __init__(self):
        print(self.x)

X()
```
The result is:

```python-traceback
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
  File "<stdin>", line 5, in __init__
AttributeError: x
```